3.6 regression: `error TS2321: Excessive stack depth comparing types 'Function<?>' and 'Function<?>'.`

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.6.2

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** excessive stack depth regression comparing types function lodash 

**Code**

`tsconfig.json`:

``` jsonc
{
    "compilerOptions": {
        "allowSyntheticDefaultImports": true,

        // Disable this and you get more information about the errors?
        "strictNullChecks": true
    }
}
```

`package.json`:

```json
{
  "dependencies": {
    "@types/lodash": "4.14.129",
    "typescript": "3.6.2"
  }
}
```

`main.ts`:

```ts
import negate from 'lodash/negate';
```

**Expected behavior:**

Compiles

**Actual behavior:**

```sh-session
$ tsc
error TS2321: Excessive stack depth comparing types 'Function<?>' and 'Function<?>'.


Found 1 error.
```

This error doesn't give any helpful information for debugging.

Upgrading `@types/lodash` to the latest version (4.14.138 at time of writing) seems to fix the issue.

If we disable `strictNullChecks`, we get a bit more information about the error (for some reason?):

```sh-session
$ tsc
error TS2321: Excessive stack depth comparing types 'Function<?>' and 'Function<?>'.

node_modules/@types/lodash/ts3.1/common/common.d.ts:194:15 - error TS2589: Type instantiation is excessively deep and possibly infinite.

194     interface Object<T> extends LoDashImplicitWrapper<T> {
                  ~~~~~~


Found 2 errors.
```

This did not error in TS 3.5, so I think it's a regression?

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

Uprading from 3.5.3 to 3.6.2 gives me this error:

`error TS2321: Excessive stack depth comparing types 'Function' and '{ [K in keyof T]?: Morphism<T[K], T[K]> | { [K in keyof T[K]]?: Morphism<T[K][K], T[K][K]> | { [K in keyof T[K][K]]?: Morphism<T[K][K][K], T[K][K][K]> | { [K in keyof T[K][K][K]]?: Morphism<T[K][K][K][K], T[K][K][K][K]> | { [K in keyof T[K][K][K][K]]?: Morphism<T[K][K][K][K][K], T[K][K][K][K][K]> | { [K in keyof T[K][K][K][K][K]]?: Morphism<T[K][K][K][K][K][K], T[K][K][K][K][K][K]> | { [K in keyof T[K][K][K][K][K][K]]?: Morphism<T[K][K][K][K][K][K][K], T[K][K][K][K][K][K][K]> | { [K in keyof T[K][K][K][K][K][K][K]]?: Morphism<T[K][K][K][K][K][K][K][K], T[K][K][K][K][K][K][K][K]> | { [K in keyof T[K][K][K][K][K][K][K][K]]?: Morphism<T[K][K][K][K][K][K][K][K][K], T[K][K][K][K][K][K][K][K][K]> | { [K in keyof T[K][K][K][K][K][K][K][K][K]]?: Morphism<T[K][K][K][K][K][K][K][K][K][K], T[K][K][K][K][K][K][K][K][K][K]> | { [K in keyof T[K][K][K][K][K][K][K][K][K][K]]?: Morphism<T[K][K][K][K][K][K][K][K][K][K][K], T[K][K][K][K][K][K][K][K][K][K][K]> | any | undefined; } | undefined; } | undefined; } | undefined; } | undefined; } | undefined; } | undefined; } | undefined; } | undefined; } | undefined; } | undefined; }'.`

Unfortunately it does not give me any hint on which dependency is causing this.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -14549,6 +14549,9 @@ namespace ts {
         // though highly unlikely, for this test to be true in a situation where a chain of instantiations is not infinitely
         // expanding. Effectively, we will generate a false positive when two types are structurally equal to at least 5
         // levels, but unequal at some level beyond that.
+        // In addition, this will also detect when an indexed access has been chained off of 5 or more times (which is essentially
+        // the dual of the structural comparison), and likewise mark the type as deeply nested, potentially adding false positives
+        // for finite but deeply expanding indexed accesses (eg, for `Q[P1][P2][P3][P4][P5]`).
         function isDeeplyNestedType(type: Type, stack: Type[], depth: number): boolean {
             // We track all object types that have an associated symbol (representing the origin of the type)
             if (depth >= 5 && type.flags & TypeFlags.Object) {
@@ -14564,7 +14567,29 @@ namespace ts {
                     }
                 }
             }
+            if (depth >= 5 && type.flags & TypeFlags.IndexedAccess) {
+                const root = getRootObjectTypeFromIndexedAccessChain(type);
+                let count = 0;
+                for (let i = 0; i < depth; i++) {
+                    const t = stack[i];
+                    if (getRootObjectTypeFromIndexedAccessChain(t) === root) {
+                        count++;
+                        if (count >= 5) return true;
+                    }
+                }
+            }
             return false;
+        }
+
+        /**
+         * Gets the leftmost object type in a chain of indexed accesses, eg, in A[P][Q], returns A
+         */
+        function getRootObjectTypeFromIndexedAccessChain(type: Type) {
+            let t = type;
+            while (t.flags & TypeFlags.IndexedAccess) {
+                t = (t as IndexedAccessType).objectType;
+            }
+            return t;
         }
 
         function isPropertyIdenticalTo(sourceProp: Symbol, targetProp: Symbol): boolean {

--- a/tests/baselines/reference/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.js
+++ b/tests/baselines/reference/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.js
@@ -1,0 +1,17 @@
+//// [comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts]
+type PartialDeep<T> = {[K in keyof T]?: PartialDeep<T[K]>};
+type Many<T> = T | readonly T[];
+
+interface Collection<T> {
+    sortBy(...iteratees: Many<PartialDeep<T>>[]): Collection<T>;
+}
+
+const x: Collection<{x: number}> = (null as any as Collection<{x: number, y: number}>);
+
+export {};
+
+
+//// [comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.js]
+"use strict";
+exports.__esModule = true;
+var x = null;

--- a/tests/baselines/reference/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.symbols
+++ b/tests/baselines/reference/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.symbols
@@ -1,0 +1,40 @@
+=== tests/cases/compiler/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts ===
+type PartialDeep<T> = {[K in keyof T]?: PartialDeep<T[K]>};
+>PartialDeep : Symbol(PartialDeep, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 0))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 17))
+>K : Symbol(K, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 24))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 17))
+>PartialDeep : Symbol(PartialDeep, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 0))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 17))
+>K : Symbol(K, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 24))
+
+type Many<T> = T | readonly T[];
+>Many : Symbol(Many, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 59))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 10))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 10))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 10))
+
+interface Collection<T> {
+>Collection : Symbol(Collection, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 32))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 3, 21))
+
+    sortBy(...iteratees: Many<PartialDeep<T>>[]): Collection<T>;
+>sortBy : Symbol(Collection.sortBy, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 3, 25))
+>iteratees : Symbol(iteratees, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 4, 11))
+>Many : Symbol(Many, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 59))
+>PartialDeep : Symbol(PartialDeep, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 0, 0))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 3, 21))
+>Collection : Symbol(Collection, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 32))
+>T : Symbol(T, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 3, 21))
+}
+
+const x: Collection<{x: number}> = (null as any as Collection<{x: number, y: number}>);
+>x : Symbol(x, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 7, 5))
+>Collection : Symbol(Collection, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 32))
+>x : Symbol(x, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 7, 21))
+>Collection : Symbol(Collection, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 1, 32))
+>x : Symbol(x, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 7, 63))
+>y : Symbol(y, Decl(comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts, 7, 73))
+
+export {};
+

--- a/tests/baselines/reference/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.types
+++ b/tests/baselines/reference/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.types
@@ -1,0 +1,25 @@
+=== tests/cases/compiler/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts ===
+type PartialDeep<T> = {[K in keyof T]?: PartialDeep<T[K]>};
+>PartialDeep : PartialDeep<T>
+
+type Many<T> = T | readonly T[];
+>Many : Many<T>
+
+interface Collection<T> {
+    sortBy(...iteratees: Many<PartialDeep<T>>[]): Collection<T>;
+>sortBy : (...iteratees: Many<PartialDeep<T>>[]) => Collection<T>
+>iteratees : Many<PartialDeep<T>>[]
+}
+
+const x: Collection<{x: number}> = (null as any as Collection<{x: number, y: number}>);
+>x : Collection<{ x: number; }>
+>x : number
+>(null as any as Collection<{x: number, y: number}>) : Collection<{ x: number; y: number; }>
+>null as any as Collection<{x: number, y: number}> : Collection<{ x: number; y: number; }>
+>null as any : any
+>null : null
+>x : number
+>y : number
+
+export {};
+

--- a/tests/cases/compiler/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts
+++ b/tests/cases/compiler/comparisonOfPartialDeepAndIndexedAccessTerminatesWithoutError.ts
@@ -1,0 +1,11 @@
+// @strict: true
+type PartialDeep<T> = {[K in keyof T]?: PartialDeep<T[K]>};
+type Many<T> = T | readonly T[];
+
+interface Collection<T> {
+    sortBy(...iteratees: Many<PartialDeep<T>>[]): Collection<T>;
+}
+
+const x: Collection<{x: number}> = (null as any as Collection<{x: number, y: number}>);
+
+export {};

